### PR TITLE
coprocessor: differentiate real/int/decimal for `not` expression

### PR DESCRIPTION
**Pull request by eurekaka**
Signed-off-by: Kenan Yao <cauchy1992@gmail.com>

## What have you changed? (mandatory)

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:
- Summarize your change (mandatory)

Add separate implementations of unary not for Real and Decimal types. The detailed reasons of this change can be found in https://github.com/pingcap/tidb/pull/10498

## What are the type of the changes? (mandatory)

- Bug fix (change which fixes an issue)

## How has this PR been tested? (mandatory)

Unit test

## Does this PR affect documentation (docs) or release note? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Refer to a related PR or issue link (optional)
PR for changes in TiDB: https://github.com/pingcap/tidb/pull/10498
PR for changes in TiPB: https://github.com/pingcap/tipb/pull/122

### CODE REVIEW

File: src/coprocessor/dag/rpn_expr/impl_op.rs [hunk 1]
```diff
@@ -126,18 +138,38 @@ mod tests {
     #[test]
     fn test_unary_not() {
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_op.rs]: Prefer to separate into test_unary_not_int, test_unary_not_real, etc.

**Comment by eurekaka** [comment]
@breeswish thanks for the comment, you are right, I will change UnaryNot to UnaryNotInt in tipb for backward compatibility later after https://github.com/tikv/tikv/issues/4864 is resolved.

File: components/tidb_query/src/rpn_expr/impl_op.rs [hunk 2]
```diff
@@ -124,20 +136,72 @@ mod tests {
...
-                .push_param(arg)
-                .evaluate(ScalarFuncSig::UnaryNot)
+                .push_param(arg.clone())
+                .evaluate(ScalarFuncSig::UnaryNotInt)
                 .unwrap();
-            assert_eq!(output, expect_output);
+            assert_eq!(
+                output,
+                expect_output,
+                "{:?}, {:?}",
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_op.rs]: `..., "{:?}", arg);` is enough

File: components/tidb_query/src/rpn_expr/impl_op.rs [hunk 3]
```diff
@@ -124,20 +136,72 @@ mod tests {
...
+        ];
+        for (arg, expect_output) in test_cases {
+            let output = RpnFnScalarEvaluator::new()
+                .push_param(arg.clone())
+                .evaluate(ScalarFuncSig::UnaryNotReal)
+                .unwrap();
+            assert_eq!(
+                output,
+                expect_output,
+                "{:?}, {:?}",
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_op.rs]: ditto.

File: components/tidb_query/src/rpn_expr/impl_op.rs [hunk 4]
```diff
@@ -124,20 +136,72 @@ mod tests {
...
+        ];
+        for (arg, expect_output) in test_cases {
+            let output = RpnFnScalarEvaluator::new()
+                .push_param(arg.clone())
+                .evaluate(ScalarFuncSig::UnaryNotDecimal)
+                .unwrap();
+            assert_eq!(
+                output,
+                expect_output,
+                "{:?}, {:?}",
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_op.rs]: ditto.

**Comment by lonng** [comment]
@eurekaka Please fix the CI failure.

**Comment by eurekaka** [comment]
@breeswish @lonng PTAL

**Comment by eurekaka** [comment]
@lonng @lonng Conflicts resolved, PTAL

**Pull request merged**